File: leetcode/101-symmetric-tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode *root) {
        return root == nullptr ? true : isMirror(root->left, root->right);
    }

    bool isMirror(TreeNode *r1, TreeNode *r2) {
        if (r1 == nullptr && r2 == nullptr) {
            return true;
        }
        if (r1 == nullptr || r2 == nullptr) {
            return false;
        }
        return r1->val == r2->val && isMirror(r1->left, r2->right)
            && isMirror(r1->right, r2->left);
    }
};

File: leetcode/101-symmetric-tree.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Boolean}
def is_symmetric(root)
  root.nil? or _is_symmetric(root.left, root.right)
end

def _is_symmetric(a, b)
  if a.nil? and b.nil?
    true
  elsif a.nil? or b.nil?
    false
  else
    a.val == b.val and _is_symmetric(a.left, b.right) and _is_symmetric(a.right, b.left)
  end
end
